Don't mount /dev/ inside privileged containers running systemd

dcermak opened the pull request:
According to https://systemd.io/CONTAINER_INTERFACE/, systemd will try take control over /dev/tty if exported, which can cause conflicts with the host's tty in privileged containers. Thus we will not expose these to privileged containers in systemd mode, as this is a bad idea according to systemd's maintainers.

This fixes https://github.com/containers/podman/issues/15878

#### Does this PR introduce a user-facing change?


```release-note
- don't expose `/dev/tty*` to privileged containers in systemd mode
```

rhatdan commented:
Please repush your PR.  I added [NO NEW TESTS NEEDED]
/approve
LGTM
@giuseppe PTAL

dcermak commented:
> Please repush your PR. I added [NO NEW TESTS NEEDED]

I have removed that and added a bats test, because my initial fix was actually broken (it got late yesterday…).

edsantiago commented on test/system/030-run.bats:
```diff
@@ -901,4 +901,24 @@ $IMAGE--c_ok" \
+    TTYs=$(ls /dev/tty*|sed '/^\/dev\/tty$/d')
+
+    if [[ $TTYs = "" ]]; then
+        false
```
Courtesy alert: it looks like you're used to regular BATS. Regular BATS is horrible for end users; this is not regular BATS. We provide mechanisms to actually make debugging possible. In this case, please replace `false` with:
```bash
    die "Did not find any /dev/ttyN devices on local host"
```
(or something else appropriate)

edsantiago commented on test/system/030-run.bats:
```diff
@@ -901,4 +901,24 @@ $IMAGE--c_ok" \
+    else
+        run_podman exec "$ctr_name" ls /dev/
+        run_output=$output
+        [[ $(echo "$run_output" | grep -c "tty") = "1"  ]]
+        [[ $(echo "$run_output" | grep "tty") = "tty"  ]]
```
Courtesy alert: again, this is regular-BATS stuff, which is impossible for anyone to debug on failure. In our environment, one correct way to do that is:
```bash
   assert "$(grep tty <<<$output)" = "tty" "No /dev/ttyN devices in container"
```
(no need to define `run_output`, it's not a useful variable). (And no need for the `-c` test either)

Reason: if this test fails, `assert` will give an error of the form "this is what I expected, and this is what I got". With regular BATS, you just get "meh, test failed but I'm not going to give you any useful information about why".

See  https://github.com/containers/podman/blob/main/test/system/README.md

edsantiago reviewed (changes requested):
Please squash your commits - no need for separate code + test commits. And please update your commit message to reflect the new state of the code. A few test suggestions inline.

Thank you, and thanks for contributing!

dcermak commented:
@edsantiago Thanks for the review, I've squashed the commits and applied your suggested changes.

edsantiago commented on test/system/030-run.bats:
```diff
@@ -901,4 +901,22 @@ $IMAGE--c_ok" \
+
+    if [[ $TTYs = "" ]]; then
+        die "Did not find any /dev/ttyN devices on local host"
+    else
```
It's a little bit confusing to use `else` after `die`, since by definition `die` terminates execution... but not worth re-pushing for, especially given our flaky CI today. I'll fix it in one of my periodic cleanup PRs.

LGTM, but I'll let other more systemd-savvy team members give final approval. Thanks again!

mheon commented:
/lgtm
/approve
